### PR TITLE
add interface to integrate Coming with Repairnator

### PR DESCRIPTION
**Pull request by SophieHYe**
In this commit, we added a new method RepairnatorFeatures.java as interface to integrate with Repairnator.

The method getLabel take input as a folder path of a pair of src and target files, and output a string label for this diff.
We use Xgboost4J, which is the java version of Xgboost for training and predicting. 

Please note that in this version:
(1) We consider libsvm fomat of features instead of csv version, for the use of Xgboost4J.
(2) We consider 150 code description features, and 26 repair features for training and testing, instead of 202 features. Because we cannot reuse the trained model for prediction with string type features as the number of features will change according to different testing string values (due to the one-hot encoding). In addition, libsvm does not seem to well support string value features.
(3) The trained model is usable but not fine turned. This model could be easily trained when more dataset considered, with the function trainModel.

### CODE REVIEW

**Comment by martinezmatias** [comment]
Hi Ye, 
Thanks a lot for implementing this feature! It's really cool!
I review the code :)

File: src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+			return UNKNOWN;
+		}
+		//Load test data in libsvm format
+		File tempFile = File.createTempFile("test", ".txt");
+		try (FileWriter sb = new FileWriter(tempFile)) {
+			sb.append(features);
+		}
+		//Load model
+		DMatrix dtest = new DMatrix(tempFile.getAbsolutePath());
+		String odsModel = this.getClass().getClassLoader().getResource("ODSTraining/ODSmodel.bin").getPath();
```
Comment (martinezmatias) [inline src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java:55]: It would be nice if the can removed the hardcoded path to the trained mode. 
Can we pass it as:
1) as method parameter? 
2) as constructor parameter?
3) using Configuration properties mechanism?  (see an example of access to config properties [here](https://github.com/SpoonLabs/coming/blob/master/src/main/java/fr/inria/coming/codefeatures/FeatureAnalyzer.java#L96))?

WDYT?

File: src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		
+		String odsModel = this.getClass().getClassLoader().getResource("ODSTraining/ODSmodel.bin").getPath();  
+
+		String trainingFile = this.getClass().getClassLoader().getResource("ODSTraining/Training-libsvm.txt").getPath();  
+
+		DMatrix trainMat = new DMatrix(trainingFile);
+		
+		Map<String, Object> params = new HashMap<String, Object>() {
+			  {
+			    put("eta", 0.3);
```
Comment (martinezmatias) [inline src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java:87]: Minor: We can move those hardcoded values to Configuration properties, so then we can play with other values.

File: src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+			  }
+			};
+			
+		Map<String, DMatrix> watches = new HashMap<String, DMatrix>() {
+				{
+				    put("train", trainMat);
+		      }
+		};		
+							
+		int nround = 20;
```
Comment (martinezmatias) [inline src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java:101]: similarly to the previous hardcoded values from `param`, it's an hyperparameter. I'd move to Configuration Properties to play with.

File: src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java (hunk 1 above)
Comment (SophieHYe) [reply src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java:55]: Sure, I will give an update in next commit. I failed to push another file which cause the build failure. Will update it as well.

File: src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java (hunk 2 above)
Comment (SophieHYe) [reply src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java:87]: No problem 👍

File: src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java (hunk 3 above)
Comment (SophieHYe) [reply src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java:101]: No problem 👍

File: src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+		String trainingFile = this.getClass().getClassLoader().getResource("ODSTraining/Training-libsvm.txt").getPath();  
+
+		DMatrix trainMat = new DMatrix(trainingFile);
+		
+		Map<String, Object> params = new HashMap<String, Object>() {
+			  {
+			    put("eta", 0.3);
+			    put("max_depth", 6);
+			    put("gamma", 1);
+			    put("objective", "binary:logistic");
```
Comment (martinezmatias) [inline src/main/java/fr/inria/coming/codefeatures/RepairnatorFeatures.java:90]: if we add to config properties, we can put a prefix to them to avoid missing with other hyperparameters. Example, `xgboost_gamma`, `xgboost_eta`

**Comment by martinezmatias** [comment]
Hi Ye

> Because we cannot reuse the trained model for prediction with string type features as the number of features will change according to different testing string values (due to the one-hot encoding)

I see your point. If I correctly remember, some of those string values are the type of the elements involved in the patch (i.e., `CtInvocation` or `CtIf` (e.g. [here](https://github.com/kth-tcs/overfitting-analysis/blob/master/StaticODS/ODSFeatures/DRR/S4R/S4Rfeatures_patch-Closure-115-ssFix-plausible.json#L9))). I agree that the set of values could differ from the training set to the testing set. However, we know all the possibles values (They can be any element from the spoon metamodel). I wonder if we can create a prefixed-size vector V, which length is the number of Spoon's classes that extends from `CtElement`.
Similarly to the ADD's feature `operator` (e.g. [here](https://github.com/kth-tcs/overfitting-analysis/blob/master/StaticODS/ODSFeatures/DRR/S4R/S4Rfeatures_patch-Closure-115-ssFix-plausible.json#L140))

**Comment by SophieHYe** [comment]
Hi Matias,

I will do the feasible analysis and add the S4R features. How about we create a new issue for that?

Currently, we will more priorities to release a maven snapshot version to the central maven repository, and test the interface with repairnator. 
Since now we have put all the logics in the coming side, we can update coming later with more features. WDYT?

**Comment by martinezmatias** [comment]
Hi Ye

> I will do the feasible analysis and add the S4R features. How about we create a new issue for that?

Done: https://github.com/SpoonLabs/coming/issues/237

> Currently, we will more priorities to release a maven snapshot version to the central maven repository, and test the interface with repairnator.

It makes perfectly sense.

> Since now we have put all the logics in the coming side, we can update coming later with more features. WDYT?

Perfect. It's fine for me.

Thanks!

**Comment by martinezmatias** [comment]
Hi Ye,
Ready to merge?

**Comment by SophieHYe** [comment]
Hi Matias,
Yes, please merge! Thanks.

**Comment by martinezmatias** [comment]
Merged! 
Thanks @SophieHYe ! Nice work!
